src/nupic/algorithms/AnomalyLikelihood.cpp

```diff
@@ -0,0 +1,340 @@
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * ---------------------------------------------------------------------
+ */
+
+#include <nupic/algorithms/AnomalyLikelihood.hpp>
+
+#include <iostream>
+#include <numeric> //accumulate, inner_product
+
+#include <nupic/utils/Log.hpp> // NTA_CHECK
+
+using namespace std;
+using namespace nupic;
+using namespace nupic::util;
+using namespace nupic::algorithms::anomaly;
+
+namespace nupic {
+  namespace algorithms {
+    namespace anomaly {
+
+      Real compute_mean(vector<Real> v); //forward declaration
+      Real compute_var(vector<Real> v, Real mean); 
+   static std::vector<Real> circularBufferToVector(boost::circular_buffer<Real> cb); //TODO replace with SlidingWindow
+   static UInt calcSkipRecords_(UInt numIngested, UInt windowSize, UInt learningPeriod);
+
+
+AnomalyLikelihood::AnomalyLikelihood(UInt learningPeriod, UInt estimationSamples, UInt historicWindowSize, UInt reestimationPeriod, UInt aggregationWindow) :
+    learningPeriod_(learningPeriod),
+    reestimationPeriod_(reestimationPeriod),
+    probationaryPeriod_(learningPeriod_+estimationSamples),
+    averagedAnomaly_(aggregationWindow) {
+        iteration_ = 0;
+        NTA_CHECK(historicWindowSize >= estimationSamples); // cerr << "estimationSamples exceeds historicWindowSize";
+        NTA_CHECK(aggregationWindow < reestimationPeriod && reestimationPeriod < historicWindowSize);
+        
+        runningAverageAnomalies_.set_capacity(historicWindowSize); 
+        runningLikelihoods_.set_capacity(historicWindowSize);
+        runningRawAnomalyScores_.set_capacity(historicWindowSize);
+        NTA_CHECK(runningLikelihoods_.capacity() == historicWindowSize);
+    }
+
+    
+Real AnomalyLikelihood::anomalyProbability(Real anomalyScore, int timestamp) {  //FIXME even timestamp is not really used, remove too? 
+    Real likelihood = DEFAULT_ANOMALY;
+
+    if (timestamp == -1) {
+      timestamp = this->iteration_;
+    }
+
+    // store into relevant variables
+    this->runningRawAnomalyScores_.push_back(anomalyScore); 
+    auto newAvg = this->averagedAnomaly_.compute(anomalyScore); 
+    this->runningAverageAnomalies_.push_back(newAvg);
+    this->iteration_++;
+    this->runningLikelihoods_.push_back(likelihood);
+    
+    // We ignore the first probationaryPeriod data points - as we cannot reliably compute distribution statistics for estimating likelihood
+    if (this->iteration_ < this->probationaryPeriod_) {
+      return DEFAULT_ANOMALY;
+    } //else {
+
+    auto anomalies = circularBufferToVector(this->runningAverageAnomalies_); 
+    
+      // On a rolling basis we re-estimate the distribution
+      if ( this->iteration_ == 0 || (this->iteration_ % reestimationPeriod_) == 0  || distribution_.name == "unknown" ) {
```

Comment:
Actually, I'm starting to like this approach. It would allow us to treat the parameters (`reestimationPeriod`, etc) as time (in units of timestamp) if `timestamp` is provided; or as iterations if it's not provided (current default).